Cargo should not scan '.*' directories for addtional Cargo.toml files

`quilt` creates a directory `.pc` under which files that are modified by patches applied are stored.

For example:
 - `curl-sys/.pc/0001-curl-sys-avoid-explicitly-linking-in-openssl-If-it-.patch/curl-sys/Cargo.toml`

When these directories are included via (for example) the .cargo/config::paths mechanism, cargo finds the Cargo.toml used for internal tracking in quilt and complains:

```
| touch target/snapshot/bin/cargo
| "/home/cody/obj/y/tmp/sysroots/x86_64-linux/usr/bin/rustc" -v
| rustc 0.13.0-dev (c637cab85 2014-11-24 02:56:35 +0000)
| target/snapshot/bin/cargo build --target x86_64-linux  --verbose
| Unable to update file:///home/cody/g/curl-rust
| 
| Caused by:
|   either a [lib] or [[bin]] section must be present
| Makefile:77: recipe for target 'cargo-x86_64-linux' failed
| make: *** [cargo-x86_64-linux] Error 101
```

--- a/src/cargo/ops/cargo_read_manifest.rs
+++ b/src/cargo/ops/cargo_read_manifest.rs
@@ -43,8 +43,9 @@ pub fn read_packages(path: &Path, source_id: &SourceId, config: &Config)
     try!(walk(path, &mut |dir| {
         trace!("looking for child package: {}", dir.display());
 
-        // Don't recurse into git databases
-        if dir.file_name().and_then(|s| s.to_str()) == Some(".git") {
+        // Don't recurse into hidden/dot directories
+        let name = dir.file_name().and_then(|s| s.to_str());
+        if name.map(|s| s.starts_with(".")) == Some(true) {
             return Ok(false)
         }
 

--- a/src/cargo/sources/path.rs
+++ b/src/cargo/sources/path.rs
@@ -255,11 +255,16 @@ impl<'cfg> PathSource<'cfg> {
         }
         for dir in try!(fs::read_dir(path)) {
             let dir = try!(dir).path();
-            match (is_root, dir.file_name().and_then(|s| s.to_str())) {
-                (_,    Some(".git")) |
-                (true, Some("target")) |
-                (true, Some("Cargo.lock")) => continue,
-                _ => {}
+            let name = dir.file_name().and_then(|s| s.to_str());
+            // Skip dotfile directories
+            if name.map(|s| s.starts_with(".")) == Some(true) {
+                continue
+            } else if is_root {
+                // Skip cargo artifacts
+                match name {
+                    Some("target") | Some("Cargo.lock") => continue,
+                    _ => {}
+                }
             }
             try!(PathSource::walk(&dir, ret, false, filter));
         }

--- a/tests/test_cargo_compile.rs
+++ b/tests/test_cargo_compile.rs
@@ -1828,6 +1828,24 @@ test!(ignore_dotfile {
                 execs().with_status(0));
 });
 
+test!(ignore_dotdirs {
+    let p = project("foo")
+        .file("Cargo.toml", r#"
+            [package]
+            name = "foo"
+            version = "0.0.1"
+            authors = []
+        "#)
+        .file("src/bin/a.rs", "fn main() {}")
+        .file(".git/Cargo.toml", "")
+        .file(".pc/dummy-fix.patch/Cargo.toml", "");
+    p.build();
+
+    assert_that(p.cargo("build"),
+                execs().with_status(0));
+});
+
+
 test!(custom_target_dir {
     let p = project("foo")
         .file("Cargo.toml", r#"

